Bug in view::take

The following does not compile with our `view::take` but does compile with `ranges::view::take`:

```cpp
#include <range/v3/numeric/accumulate.hpp>   // ranges::accumulate
#include <range/v3/view/take.hpp>

#include <seqan3/io/sequence_file/all.hpp>
#include <seqan3/io/stream/debug_stream.hpp>
#include <seqan3/range/view/take.hpp>
#include <seqan3/std/ranges>

using namespace seqan3;

int main()
{
    sequence_file_input fin{"/tmp/my.fastq"};

    auto minimum_quality_filter = std::view::filter([] (auto const & rec)
    {
        auto qual = get<field::QUAL>(rec) | std::view::transform([] (auto q) { return q.to_phred(); });
        double sum = ranges::accumulate(qual.begin(), qual.end(), 0);
        return sum/std::ranges::size(qual) >= 40;
    });

    for (auto & rec : fin | minimum_quality_filter | view::take(2))
    {
        debug_stream << "ID: " << get<field::ID>(rec) << '\n';
    }
}
```
Error message:
<summary><details>

```
In file included from /srv/public/smehringer/seqan3/include/seqan3/range/view/take.hpp:25,
                 from /srv/public/smehringer/seqan3/include/seqan3/io/sequence_file/format_fasta.hpp:39,
                 from /srv/public/smehringer/seqan3/include/seqan3/io/sequence_file/all.hpp:20,
                 from /srv/public/smehringer/seqan3/doc/tutorial/sequence_file/solution3.cpp:43:
/srv/public/smehringer/seqan3/include/seqan3/range/detail/inherited_iterator_base.hpp: In instantiation of ‘class seqan3::detail::inherited_iterator_base<seqan3::detail::view_take<ranges::remove_if_view<ranges::ref_view<seqan3::sequence_file_input<> >, ranges::logical_negate_<main()::<lambda(const auto:16&)> > >, false, false>::iterator_type<ranges::remove_if_view<ranges::ref_view<seqan3::sequence_file_input<> >, ranges::logical_negate_<main()::<lambda(const auto:16&)> > > >, ranges::basic_iterator<ranges::adaptor_cursor<seqan3::detail::in_file_iterator<seqan3::sequence_file_input<> >, ranges::remove_if_view<ranges::ref_view<seqan3::sequence_file_input<> >, ranges::logical_negate_<main()::<lambda(const auto:16&)> > >::adaptor> > >’:
/srv/public/smehringer/seqan3/include/seqan3/range/view/take.hpp:68:11:   required from ‘class seqan3::detail::view_take<ranges::remove_if_view<ranges::ref_view<seqan3::sequence_file_input<> >, ranges::logical_negate_<main()::<lambda(const auto:16&)> > >, false, false>::iterator_type<ranges::remove_if_view<ranges::ref_view<seqan3::sequence_file_input<> >, ranges::logical_negate_<main()::<lambda(const auto:16&)> > > >’
/srv/public/smehringer/seqan3/doc/tutorial/sequence_file/solution3.cpp:61:66:   required from here
/srv/public/smehringer/seqan3/include/seqan3/range/detail/inherited_iterator_base.hpp:56:91: error: no type named ‘iterator_category’ in ‘struct std::iterator_traits<ranges::basic_iterator<ranges::adaptor_cursor<seqan3::detail::in_file_iterator<seqan3::sequence_file_input<> >, ranges::remove_if_view<ranges::ref_view<seqan3::sequence_file_input<> >, ranges::logical_negate_<main()::<lambda(const auto:16&)> > >::adaptor> > >’
     using iterator_category     = typename std::iterator_traits<base_t>::iterator_category;
                                                                                           ^
In file included from /srv/public/smehringer/seqan3/include/seqan3/io/sequence_file/format_fasta.hpp:39,
                 from /srv/public/smehringer/seqan3/include/seqan3/io/sequence_file/all.hpp:20,
                 from /srv/public/smehringer/seqan3/doc/tutorial/sequence_file/solution3.cpp:43:
/srv/public/smehringer/seqan3/include/seqan3/range/view/take.hpp: In instantiation of ‘class seqan3::detail::view_take<ranges::remove_if_view<ranges::ref_view<seqan3::sequence_file_input<> >, ranges::logical_negate_<main()::<lambda(const auto:16&)> > >, false, false>::iterator_type<ranges::remove_if_view<ranges::ref_view<seqan3::sequence_file_input<> >, ranges::logical_negate_<main()::<lambda(const auto:16&)> > > >’:
/srv/public/smehringer/seqan3/doc/tutorial/sequence_file/solution3.cpp:61:66:   required from here
/srv/public/smehringer/seqan3/include/seqan3/range/view/take.hpp:112:100: error: no type named ‘iterator_category’ in ‘struct std::iterator_traits<ranges::basic_iterator<ranges::adaptor_cursor<seqan3::detail::in_file_iterator<seqan3::sequence_file_input<> >, ranges::remove_if_view<ranges::ref_view<seqan3::sequence_file_input<> >, ranges::logical_negate_<main()::<lambda(const auto:16&)> > >::adaptor> > >’
         using iterator_category     = typename std::iterator_traits<base_base_t>::iterator_category;
                                                                                                    ^
In file included from /srv/public/smehringer/seqan3/submodules/range-v3/include/range/v3/iterator/basic_iterator.hpp:27,
                 from /srv/public/smehringer/seqan3/submodules/range-v3/include/range/v3/iterator/reverse_iterator.hpp:19,
                 from /srv/public/smehringer/seqan3/submodules/range-v3/include/range/v3/range/access.hpp:29,
                 from /srv/public/smehringer/seqan3/submodules/range-v3/include/range/v3/numeric/accumulate.hpp:17,
                 from /srv/public/smehringer/seqan3/doc/tutorial/sequence_file/solution3.cpp:40:
/srv/public/smehringer/seqan3/submodules/range-v3/include/range/v3/utility/semiregular.hpp:112:9: error: ‘ranges::semiregular<T>::semiregular(ranges::semiregular<T>&&) noexcept [with T = ranges::logical_negate_<main()::<lambda(const auto:16&)> >]’ used but never defined [-Werror]
         semiregular(semiregular &&that)
         ^~~~~~~~~~~
```

</details></summary>

--- a/include/seqan3/core/metafunction/iterator.hpp
+++ b/include/seqan3/core/metafunction/iterator.hpp
@@ -113,6 +113,110 @@ struct size_type<it_t>
 
 // see specialisation for ranges in core/metafunction/range.hpp
 
+// ----------------------------------------------------------------------------
+// iterator_tag
+// ----------------------------------------------------------------------------
+
+/*!\brief Type metafunction that deduces the
+ * [iterator_category](https://en.cppreference.com/w/cpp/iterator/iterator_tags) from the modelled
+ * concept [Type metafunction].
+ * \tparam it_t The type to operate on.
+ *
+ * <table>
+ *   <tr>
+ *     <th>Modelled concept</th>
+ *     <th>iterator_tag<it_t>::type</th>
+ *   </tr>
+ *   <tr>
+ *     <td>\ref std::InputIterator "std::InputIterator<it_t>"</td>
+ *     <td>std::input_iterator_tag</td>
+ *   </tr>
+ *   <tr>
+ *     <td>\ref std::InputIterator "!std::InputIterator<it_t>" &amp;&amp;
+ *         \ref std::OutputIterator "std::OutputIterator<it_t, value_type<it_t>>"</td>
+ *     <td>std::output_iterator_tag</td>
+ *   </tr>
+ *   <tr>
+ *     <td>\ref std::ForwardIterator "std::ForwardIterator<it_t>"</td>
+ *     <td>std::forward_iterator_tag</td>
+ *   </tr>
+ *   <tr>
+ *     <td>\ref std::BidirectionalIterator "std::BidirectionalIterator<it_t>"</td>
+ *     <td>std::bidirectional_iterator_tag</td>
+ *   </tr>
+ *   <tr>
+ *     <td>\ref std::RandomAccessIterator "std::RandomAccessIterator<it_t>"</td>
+ *     <td>std::random_access_iterator_tag</td>
+ *   </tr>
+ * </table>
+ *
+ * \attention
+ * If [std::iterator_traits<it_t>::iterator_category](https://en.cppreference.com/w/cpp/iterator/iterator_traits)
+ * is defined for a type `it_t`, this metafunction acts as an alias for it.
+ * If it is not defined and no concepts are modelled, iterator_tag<it_t>::type is not defined.
+ */
+template <typename it_t>
+struct iterator_tag
+{
+SEQAN3_DOXYGEN_ONLY(
+    //!\brief The [iterator_category](https://en.cppreference.com/w/cpp/iterator/iterator_tags).
+    using type = iterator_category;
+)
+};
+
+//!\cond
+template <typename it_t>
+    requires requires { typename std::iterator_traits<it_t>::iterator_category; }
+struct iterator_tag<it_t>
+{
+    using type = typename std::iterator_traits<it_t>::iterator_category;
+};
+
+template <std::InputIterator it_t>
+    requires !requires { typename std::iterator_traits<it_t>::iterator_category; }
+struct iterator_tag<it_t>
+{
+    using type = std::input_iterator_tag;
+};
+
+template <typename it_t>
+    requires !std::InputIterator<it_t> && std::OutputIterator<it_t, value_type_t<it_t>> &&
+             !requires { typename std::iterator_traits<it_t>::iterator_category; }
+struct iterator_tag<it_t>
+{
+    using type = std::output_iterator_tag;
+};
+
+template <std::ForwardIterator it_t>
+    requires !requires { typename std::iterator_traits<it_t>::iterator_category; }
+struct iterator_tag<it_t>
+{
+    using type = std::forward_iterator_tag;
+};
+
+template <std::BidirectionalIterator it_t>
+    requires !requires { typename std::iterator_traits<it_t>::iterator_category; }
+struct iterator_tag<it_t>
+{
+    using type = std::bidirectional_iterator_tag;
+};
+
+template <std::RandomAccessIterator it_t>
+    requires !requires { typename std::iterator_traits<it_t>::iterator_category; }
+struct iterator_tag<it_t>
+{
+    using type = std::random_access_iterator_tag;
+};
+//!\endcond
+
+/*!\brief Return the `iterator_category` type of the input type [Type metafunction, shortcut].
+ * \tparam it_t The type to operate on.
+ */
+template <typename it_t>
+//!\cond
+    requires requires { typename iterator_tag<it_t>::type; }
+//!\endcond
+using iterator_tag_t = typename iterator_tag<it_t>::type;
 //!\}
 
 } // namespace seqan3

--- a/include/seqan3/range/detail/inherited_iterator_base.hpp
+++ b/include/seqan3/range/detail/inherited_iterator_base.hpp
@@ -15,6 +15,7 @@
 #include <cassert>
 #include <type_traits>
 
+#include <seqan3/core/metafunction/iterator.hpp>
 #include <seqan3/std/iterator>
 
 namespace seqan3::detail
@@ -53,7 +54,7 @@ public:
     using value_type            = typename std::iterator_traits<base_t>::value_type;
     using reference             = typename std::iterator_traits<base_t>::reference;
     using pointer               = typename std::iterator_traits<base_t>::pointer;
-    using iterator_category     = typename std::iterator_traits<base_t>::iterator_category;
+    using iterator_category     = iterator_tag_t<base_t>;
     //!\}
 
     /*!\name Constructors, destructor and assignment

--- a/include/seqan3/range/view/pairwise_combine.hpp
+++ b/include/seqan3/range/view/pairwise_combine.hpp
@@ -81,7 +81,7 @@ private:
         using value_type        = std::tuple<underlying_val_t, underlying_val_t>;
         using reference         = std::tuple<underlying_ref_t, underlying_ref_t>;
         using pointer           = void;
-        using iterator_category = typename std::iterator_traits<underlying_iterator_type>::iterator_category;
+        using iterator_category = iterator_tag_t<underlying_iterator_type>;
         //!\}
 
         /*!\name Constructors, destructor and assignment

--- a/include/seqan3/range/view/take.hpp
+++ b/include/seqan3/range/view/take.hpp
@@ -109,7 +109,7 @@ private:
         using value_type            = typename std::iterator_traits<base_base_t>::value_type;
         using reference             = typename std::iterator_traits<base_base_t>::reference;
         using pointer               = typename std::iterator_traits<base_base_t>::pointer;
-        using iterator_category     = typename std::iterator_traits<base_base_t>::iterator_category;
+        using iterator_category     = iterator_tag_t<base_base_t>;
         //!\}
 
         /*!\name Arithmetic operators

--- a/include/seqan3/range/view/take_line.hpp
+++ b/include/seqan3/range/view/take_line.hpp
@@ -96,7 +96,7 @@ private:
         using value_type            = typename std::iterator_traits<base_base_t>::value_type;
         using reference             = typename std::iterator_traits<base_base_t>::reference;
         using pointer               = typename std::iterator_traits<base_base_t>::pointer;
-        using iterator_category     = typename std::iterator_traits<base_base_t>::iterator_category;
+        using iterator_category     = iterator_tag_t<base_base_t>;
         //!\}
 
         /*!\name Arithmetic operators

--- a/include/seqan3/range/view/take_until.hpp
+++ b/include/seqan3/range/view/take_until.hpp
@@ -121,7 +121,7 @@ private:
         using value_type            = typename std::iterator_traits<base_base_t>::value_type;
         using reference             = typename std::iterator_traits<base_base_t>::reference;
         using pointer               = typename std::iterator_traits<base_base_t>::pointer;
-        using iterator_category     = typename std::iterator_traits<base_base_t>::iterator_category;
+        using iterator_category     = iterator_tag_t<base_base_t>;
         //!\}
 
         /*!\name Comparison operators

